homeassistant/components/lifx/__init__.py

```diff
@@ -22,39 +45,173 @@
+            "LIFX starting discovery with interval: %s and migrating: %s",
+            discovery_interval,
+            self.migrating,
+        )
+        self._cancel_discovery = async_track_time_interval(
+            self.hass, self.async_discovery, discovery_interval
+        )
+
+    async def async_discovery(self, *_: Any) -> None:
+        """Discovery and migrate LIFX devics."""
+        migrating_was_in_progress = self.migrating
+
+        async with self.lock:
+            discovered = await async_discover_devices(self.hass)
+
+            if legacy_entry := async_get_legacy_entry(self.hass):
+                migration_complete = await async_legacy_migration(
+                    self.hass, legacy_entry, discovered
+                )
+                if migration_complete and migrating_was_in_progress:
+                    self.migrating = False
+                    _LOGGER.debug(
+                        "LIFX migration complete, switching to normal discovery interval: %s",
+                        DISCOVERY_INTERVAL,
+                    )
+                    self.async_setup_discovery_interval()
+
+            if discovered:
+                async_trigger_discovery(self.hass, discovered)
+
+
+async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
+    """Set up the LIFX component."""
+    hass.data[DOMAIN] = {}
+    migrating = bool(async_get_legacy_entry(hass))
+    discovery_manager = LIFXDiscoveryManager(hass, migrating)
+
+    @callback
+    def _async_delayed_discovery(now: datetime) -> None:
+        """Start an untracked task to discover devices.
+
+        We do not want the discovery task to block startup.
+        """
+        asyncio.create_task(discovery_manager.async_discovery())
+
+    # Let the system settle a bit before starting discovery
+    # to reduce the risk we miss devices because the event
+    # loop is blocked at startup.
+    discovery_manager.async_setup_discovery_interval()
+    async_call_later(hass, DISCOVERY_COOLDOWN, _async_delayed_discovery)
+    hass.bus.async_listen_once(
+        EVENT_HOMEASSISTANT_STARTED, discovery_manager.async_discovery
+    )
 
     return True
 
 
 async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
     """Set up LIFX from a config entry."""
+    if async_entry_is_legacy(entry):
+        return True
+
+    if legacy_entry := async_get_legacy_entry(hass):
+        # If the legacy entry still exists, harvest the entities
+        # that are moving to this config entry.
+        await async_migrate_entities_devices(hass, legacy_entry.entry_id, entry)
+
+    assert entry.unique_id is not None
+    if DATA_LIFX_MANAGER not in hass.data:
+        manager = LIFXManager(hass)
+        hass.data[DATA_LIFX_MANAGER] = manager
+        manager.async_setup()
+
+    host = entry.data[CONF_HOST]
+    connection = LIFXConnection(host, TARGET_ANY)
+    try:
+        await connection.async_setup()
+    except socket.gaierror as ex:
```

Comment:
I assume `aiolifx` doesn't wrap this in a LIFX-specific error?

Reply:
This happens when the user ends a dns name and it cannot be resolved.  The lib doesn't handle this